tests/30rooms/12thirdpartyinvite.pl

```diff
@@ -4,377 +4,199 @@
-
-            do_request_json_for( $invitee,
-               method  => "POST",
-               uri     => "/api/v1/rooms/$room_id/join",
-               content => {
-                  token            => $token,
-                  public_key       => $public_key,
-                  key_validity_url => "https://$id_server/_matrix/identity/api/v1/pubkey/isvalid",
-                  sender           => $user->user_id,
-               }
-            );
-         });
-   };
+      my ( $inviter, $other_member, $invitee, $locations ) = @_;
+      my $hs_uribase = $locations->[0];
 
-test "3pid invite join with wrong key_validity_url are rejected",
-   requires => [ @user_preparers, qw( test_http_server_hostandport )],
+      start_id_server()->then( sub {
+         my ( $id_server ) = @_;
 
-   do => sub {
-      my ( $user, $invitee ) = @_;
-
-      make_3pid_invite(
-         inviter             => $user,
-         invitee             => $invitee,
-         expect_join_success => 0,
-         join_sub            => sub {
-            my ( $token, $public_key, $signed, $room_id, $id_server ) = @_;
-
-            do_request_json_for( $invitee,
-               method  => "POST",
-               uri     => "/api/v1/rooms/$room_id/join",
-               content => {
-                  token            => $token,
-                  public_key       => $public_key,
-                  signed           => $signed,
-                  key_validity_url => "https://wrongdoesnotexist$id_server/_matrix/identity/api/v1/pubkey/isvalid",
-                  sender           => $user->user_id,
-               }
-            );
-         });
-   };
+         my $room_id;
 
-test "3pid invite join with missing key_validity_url are rejected",
-   requires => [ @user_preparers, qw( test_http_server_hostandport )],
+         matrix_create_room( $inviter, visibility => "private" )
+         ->then(sub {
+            ( $room_id ) = @_;
 
-   do => sub {
-      my ( $user, $invitee ) = @_;
-
-      make_3pid_invite(
-         inviter             => $user,
-         invitee             => $invitee,
-         expect_join_success => 0,
-         join_sub            => sub {
-            my ( $token, $public_key, $signed, $room_id ) = @_;
-
-            do_request_json_for( $invitee,
-               method  => "POST",
-               uri     => "/api/v1/rooms/$room_id/join",
-               content => {
-                  token      => $token,
-                  public_key => $public_key,
-                  signed     => $signed,
-                  sender     => $user->user_id,
-               }
-            );
-         });
+             matrix_invite_user_to_room( $inviter, $other_member, $room_id );
+         })->then(sub {
+             matrix_join_room( $other_member, $room_id );
+         })->then(sub {
+            do_3pid_invite( $inviter, $room_id, $id_server->name, $invitee_email )
+         })->then( sub {
```

Comment:
Some consistency on `->then( sub {` vs `->then(sub {` would be nice here. Usually the former is the prevailing style around sytest.